docs: 34.0.0 release notes (#18231)
apache/druid: docs/release-info/release-notes.md

As part of this change, new Coordinator APIs are available. For more information, see [Coordinator APIs for clones](#coordinator-apis-for-clones).

[#17863](https://github.com/apache/druid/pull/17863) [#17899](https://github.com/apache/druid/pull/17899) [#17956](https://github.com/apache/druid/pull/17956) 
### Embedded kill tasks on the Overlord (Experimental)

You can now run kill tasks directly on the Overlord itself. Embedded kill tasks provide several benefits; they:

- Kill segments as soon as they're eligible 
- Don't take up tasks slot
- finish faster since they use optimized metadata queries and don't launch a new JVM
- Kill a small number of segments per task, ensuring locks on an interval aren't held for too long
- Skip locked intervals to avoid head-of-line blocking
- Require minimal configuration
- Can keep up with a large number of unused segments in the cluster

This feature is controlled by the following configs:

- `druid.manager.segments.killUnused.enabled` - Whether the feature is enabled or not

> - `druid.manager.segments.killUnused.bufferPeriod` - The amount of time that a segment must be unused before it is able to be permanently removed from metadata and deep storage. This can serve as a buffer period to prevent data loss if data ends up being needed after being marked unused.

Reviewer: Do you think it's worth mentioning what the defaults for these configs are? (false and `P30D`)

Reviewer 2: +1

> - `druid.manager.segments.killUnused.enabled` - Whether the feature is enabled or not
> - `druid.manager.segments.killUnused.bufferPeriod` - The amount of time that a segment must be unused before it is able to be permanently removed from metadata and deep storage. This can serve as a buffer period to prevent data loss if data ends up being needed after being marked unused.

Author: ```suggestion
- `druid.manager.segments.killUnused.enabled` - Whether the feature is enabled or not (Defaults to `false`)
- `druid.manager.segments.killUnused.bufferPeriod` - The amount of time that a segment must be unused before it is able to be permanently removed from metadata and deep storage. This can serve as a buffer period to prevent data loss if data ends up being needed after being marked unused (Defaults to `P30D`)
```

> T use embedded kill tasks, you need to have segment metadata cache enabled.

Reviewer: ```suggestion
To use embedded kill tasks, you need to have segment metadata cache enabled.
```

As part of this feature, [new metrics](#overlord-kill-task-metrics) have been added.

> [#18028](https://github.com/apache/druid/pull/18028)

Reviewer: ```suggestion
[#18028](https://github.com/apache/druid/pull/18028) [#18124](https://github.com/apache/druid/pull/18124)
```

### Preferred tier selection 
You can now configure the Broker service to prefer  Historicals on a specific tier. This can help ensure Druid executes queries within the same availability zone if you have Druid deployed across multiple availability zones.

[#18136](https://github.com/apache/druid/pull/18136)

### Dart improvements NEED TO WRITE

Dart specific endpoints have been removed and folded into SqlResource. [#18003](https://github.com/apache/druid/pull/18003)
Added a new engine QueryContext parameter. The value can be native or msq-dart. The value determines the engine used to run the query. The default value is native. [#18003](https://github.com/apache/druid/pull/18003)

MSQ Dart is now able to query real-time tasks by setting the query context parameter includeSegmentSource to realtime, in a similar way to MSQ tasks. [#18076](https://github.com/apache/druid/pull/18076)

### `SegmentMetadataCache` on the Coordinator

[#17996](https://github.com/apache/druid/pull/17996) [#17935](https://github.com/apache/druid/pull/17935)

## Functional area and related changes

This section contains detailed release notes separated by areas.

### Web console

#### SET statements

> The web console supports using SET statements to specify query context parameters. For example, if you include `SET timeout = 20000;` in your query, the timeout query context parameter is set. 

Author: I can move this to the headline section. This entry is specifically about the web console support for SET though. 

An example that hits the SQL endpoint, uses dart etc should probably go in the Dart section and/or the improved HTTP endpoint entry

When using this feature, make sure you set `useConcurrentLocks` to `true` for the `context` field in the supervisor spec.

[#18149](https://github.com/apache/druid/pull/18149) [#18082](https://github.com/apache/druid/pull/18082)

##### Supervisors and the underlying input stream

Seekable stream supervisors (Kafka, Kinesis, and Rabbit) can no longer be updated to ingest from a different input stream (such as a topic for Kafka). Since such a change is not fully supported by the underlying system, a request to make such a change will result in a 400 error. 

[#17955](https://github.com/apache/druid/pull/17955) [#17975](https://github.com/apache/druid/pull/17975)

##### Other streaming ingestion improvements

- Improved streaming ingestion so that it automatically determine the maximum number of columns to merge [#17917](https://github.com/apache/druid/pull/17917)

### Querying

#### Metadata query for segments

You can use a segment metadata query to find the list of projections attached to a segment.

[#18119](https://github.com/apache/druid/pull/18119) [#18223](https://github.com/apache/druid/pull/)

#### `json_merge()` improvement

> `json_merge()` is now SQL-compliant when arguments are null. The function now returns null if any argument is null. For example, queries like SELECT JSON_MERGE(null, null) and SELECT JSON_MERGE(null, '{}') will return null instead of throwing an error.
> 

Author: ```suggestion
`json_merge()` is now SQL-compliant when arguments are null. The function now returns null if any argument is null. For example, queries like SELECT JSON_MERGE(null, null) and SELECT JSON_MERGE(null, '{}') will return null instead of throwing an error.

[#17983](https://github.com/apache/druid/pull/17983)
```

- Kill segments as soon as they're eligible 
- Don't take up tasks slot
- finish faster since they use optimized metadata queries and don't launch a new JVM
- Kill a small number of segments per task, ensuring locks on an interval aren't held for too long
- Skip locked intervals to avoid head-of-line blocking
- Require minimal configuration
- Can keep up with a large number of unused segments in the cluster

This feature is controlled by the following configs:

- `druid.manager.segments.killUnused.enabled` - Whether the feature is enabled or not (Defaults to `false`)
- `druid.manager.segments.killUnused.bufferPeriod` - The amount of time that a segment must be unused before it is able to be permanently removed from metadata and deep storage. This can serve as a buffer period to prevent data loss if data ends up being needed after being marked unused (Defaults to `P30D`)

To use embedded kill tasks, you need to have segment metadata cache enabled.

As part of this feature, [new metrics](#overlord-kill-task-metrics) have been added.

[#18028](https://github.com/apache/druid/pull/18028) [#18124](https://github.com/apache/druid/pull/18124)

> ### Preferred tier selection 
> You can now configure the Broker service to prefer  Historicals on a specific tier. This can help ensure Druid executes queries within the same availability zone if you have Druid deployed across multiple availability zones.

Reviewer 2: suggestion with the PR authors updated notes https://github.com/apache/druid/pull/18136#issuecomment-3096494116

```suggestion
### Preferred tier selection 
You can now configure the Broker service to prefer Historicals on a specific tier. This is useful for across availability zone deployment. Brokers in one AZ select historicals in the same AZ by default but still keeps the ability to select historical nodes in another AZ if historicals in the same AZ are not available.

To enable, set property `druid.broker.select.tier` to `perferred` in Broker runtime properties. You can then configure `druid.broker.select.tier.preferred.tier` to the tier you want each broker to prefer (i.e. for brokers in AZ1, you could set this to the tier name of your AZ1 historical servers).
```

Reviewer 3: Catch on spelling: `perferred` -> `preferred`.

This section contains important information about new and existing features.

### Hadoop-based ingestion

> Hadoop-based ingestion has been deprecated since Druid 32.0. You must now opt-in to using the deprecated `index_hadoop` task type. If you don't do this, your Hadoop-based ingestion tasks will fail.
> 
> To opt-in, set `druid.indexer.task.allowHadoopTaskExecution` to `true` in your `common.runtime.properties` file.
> 
> [#18239](https://github.com/apache/druid/pull/18239)

Author: ```suggestion
Hadoop-based ingestion has been deprecated since Druid 32.0 and will be removed as early as Druid 35.0.0. 

We recommend one of Druid's other supported ingestion methods, such as [SQL-based ingestion](../multi-stage-query/index.md) or [MiddleManager-less ingestion using Kubernetes](../development/extensions-core/k8s-jobs.md).

As part of this change, you must now opt-in to using the deprecated `index_hadoop` task type. If you don't do this, your Hadoop-based ingestion tasks will fail.

To opt-in, set `druid.indexer.task.allowHadoopTaskExecution` to `true` in your `common.runtime.properties` file.

[#18239](https://github.com/apache/druid/pull/18239)
```

Author: How about this @capistrant ? If this is good, we can copy it to the upgrade section and the upgrade notes page

- Changed the logging level for query cancellation from `warn` to `info` to reduce noise [#18046](https://github.com/apache/druid/pull/18046)
- Changed query logging so that SQL queries that can't be parsed are no longer logged and don't emit metrics [#18102](https://github.com/apache/druid/pull/18102)
- Changed the logging level for lifecycle from `debug` to `info` [#17884](https://github.com/apache/druid/pull/17884)
- Added  `groupId` and `tasks` to Overlord logs [#17980](https://github.com/apache/druid/pull/17980)
- You can now use the `druid.request.logging.rollPeriod` to configure the log rotation period (default 1 day) [#17976](https://github.com/apache/druid/pull/17976)
- Improved metric emission on the Broker to include per-query result-level caching (`query/resultCache/hit` returning `1` means the cache was used) [#18063](https://github.com/apache/druid/pull/18063)

### Extensions

#### Kubernetes

- The task runner log now includes the task id for a job and includes a log before a job is created [#18105](https://github.com/apache/druid/pull/18105)

### Documentation improvements

## Upgrade notes and incompatible changes

### Upgrade notes

#### Hadoop-based ingestion

> Hadoop-based ingestion has been deprecated since Druid 32.0. You must now opt-in to using the deprecated `index_hadoop` task type. If you don't do this, your Hadoop-based ingestion tasks will fail.
> 
> To opt-in, set `druid.indexer.task.allowHadoopTaskExecution` to `true` in your `common.runtime.properties` file.
> 
> Hadoop-based ingestion deprecated. Use [SQL-based ingestion](../multi-stage-query/index.md) instead of MapReduce or [MiddleManager-less ingestion using Kubernetes](../development/extensions-core/k8s-jobs.md) instead of YARN.
> 
> 
> [#18239](https://github.com/apache/druid/pull/18239)

Author: ```suggestion
Hadoop-based ingestion has been deprecated since Druid 32.0 and will be removed as early as Druid 35.0.0. 

We recommend one of Druid's other supported ingestion methods, such as [SQL-based ingestion](../multi-stage-query/index.md) or [MiddleManager-less ingestion using Kubernetes](../development/extensions-core/k8s-jobs.md).

As part of this change, you must now opt-in to using the deprecated `index_hadoop` task type. If you don't do this, your Hadoop-based ingestion tasks will fail.

To opt-in, set `druid.indexer.task.allowHadoopTaskExecution` to `true` in your `common.runtime.properties` file.

[#18239](https://github.com/apache/druid/pull/18239)
```

When using this feature, make sure you set `useConcurrentLocks` to `true` for the `context` field in the supervisor spec.

[#18149](https://github.com/apache/druid/pull/18149) [#18082](https://github.com/apache/druid/pull/18082)

##### Supervisors and the underlying input stream

Seekable stream supervisors (Kafka, Kinesis, and Rabbit) can no longer be updated to ingest from a different input stream (such as a topic for Kafka). Since such a change is not fully supported by the underlying system, a request to make such a change will result in a 400 error. 

[#17955](https://github.com/apache/druid/pull/17955) [#17975](https://github.com/apache/druid/pull/17975)

##### Other streaming ingestion improvements

- Improved streaming ingestion so that it automatically determine the maximum number of columns to merge [#17917](https://github.com/apache/druid/pull/17917)

### Querying

#### Metadata query for segments

You can use a segment metadata query to find the list of projections attached to a segment.

[#18119](https://github.com/apache/druid/pull/18119)

> 

Reviewer: Looks like a release note for https://github.com/apache/druid/pull/17983  is missing

#### Other querying improvements

- You can now perform big decimal aggregations using the MSQ task engine [#18164](https://github.com/apache/druid/pull/18164)
- Changed `MV_OVERLAP` and `MV_CONTAINS` functions now aligns more closely with the native `inType` filter [#18084](https://github.com/apache/druid/pull/18084)

> - Improved query handling when some segments are missing on Historicals. Druid no longer incorrectly returns partial results [#18025](https://github.com/apache/druid/pull/18025)

Reviewer: Reworded since this is a bit more specific:
```suggestion
- Improved query handling when segments are temporarily missing on Historicals but not detected by Brokers. Druid doesn't return partial results incorrectly in such cases. [#18025](https://github.com/apache/druid/pull/18025)
```
